alternator/server.cc

```diff
@@ -211,9 +212,28 @@ class local_nodelist_handler : public gated_handler {
         // using _gossiper().get_live_members(). But getting
         // just the list of live nodes in this DC needs more elaborate code:
         auto& topology = _proxy.get_token_metadata_ptr()->get_topology();
-        sstring local_dc = topology.get_datacenter();
-        std::unordered_set<gms::inet_address> local_dc_nodes = topology.get_datacenter_endpoints().at(local_dc);
+        // /localnodes lists nodes in a single DC. By default the DC of this
+        // server is used, but it can be overriden by a "dc" query option.
+        // If the DC does not exist, we return an empty list - not an error.
```

Comment:
Why not return some informative message saying that the DC in question doesn't exist?

Reply:
BTW. do we return http return code 200 or 204 for the case where there's nothing to return?

Reply:
Good question. I deliberated myhself what to do in this case. I can return an HTTP error (not just "a message") in this case, and also if the requested rack doesn't exist in the gossiper.

But what does it mean that a DC or rack doesn't "exist"? The place I look for it is in the gossiper, so if a rack or dc doesn't exist there is just means there is no node gossiped with this dc and rack. What if the rack or DC actually does "exist" in theory, but just doesn't have any node (yet, or now, or I don't know what)? Wouldn't it make sense to return an empty list like I do?

Also, I thought that the user (as in https://github.com/scylladb/alternator-load-balancing/issues/31) will anyway look for a specific rack and if an empty list is returned, fall back to a query without a rack - and I didn't think the code cares of the rack "doesn't exist" or exists but doesn't have any live node.

But if you feel that this case should result in an HTTP error and not an empty list, I can change it.

Note that users have other ways, not through "/localnodes", to get names of racks and dcs of different nodes. They shouldn't need /localnodes to be the way to check for spelling mistakes in rack or dc names.

Reply:
> BTW. do we return http return code 200 or 204 for the case where there's nothing to return?

This API (which we invented, by the way - it doesn't exist in DynamoDB so there's no compatibility question),
returns a JSON list. So an empty list looks like the string `[]`, and uses HTTP response code 200 (OK), not something else.
By the way, even if the response was completely empty, it can use response code 200. There is no requirement in the standard to return 204 if the response is completely empty (which it isn't, anyway).

Reply:
204 isn't relevant - you can't return the string '[]' (empty list), which is a response body, and use status code 204 which says there is no body.
We can make it an error if we want, or the way I did it now (NOT an error), but 204 is not an option.

Don't forget that it's entirely possible that there will be no live node in a certain rack (in a small, 3-node DC, spanning 3 racks). You have to seriously consider if you want to report an error in this case (rack doesn't exist) or just return a list of zero nodes. I liked my solution of returning an empty list, but I'm not married to this solution.

Reply:
We need to make a decision here - is it fine that we return an empty list ("no nodes in this DC") instead of a "DC doesn't exist" error?
Maybe someone with more snitch experience (@vladzcloudius) can help us: Is there a difference between "a DC that doesn't exist" and "a DC that doesn't have any nodes"? Does Scylla know which DCs "exist" even before any nodes are added to them?